File: level2/다리를_지나는_트럭/solution.js
function solution(bridge_length, weight, truck_weights) {
  var answer = 0; // 총 경과시간
  var bridge = new Array(bridge_length).fill(0); // 다리 초기화
  var upWeight = 0; // 다리위 트럭 무게의 합
  for (var i = 0; i < truck_weights.length; i++) {
    answer++;
    upWeight -= bridge.shift();
    var w = truck_weights[i];
    // 다리위에 트럭이 올라가면서 제한 무게를 초과 할 때
    if (weight < upWeight + w) {
      bridge.push(0);
      i--;
    } else {
      bridge.push(w);
      upWeight += w;
    }
  }

  // 다리위에 남은 트럭 빼기
  while (upWeight != 0) {
    upWeight -= bridge.shift();
    answer++;
  }
  return answer;
}

console.log(solution(2, 10, [7, 4, 5, 6]));
console.log(solution(100, 100, [10]));
console.log(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]));

File: level2/다리를_지나는_트럭/solution.rb
def solution(bridge_length, weight, truck_weights)
  answer = 0
  bridge_weight = 0 # 다리 위 트럭 무게
  bridge = Array.new(bridge_length, 0) # 다리 초기화

  truck_weights.each do |tw|
    answer += 1
    bridge_weight -= bridge.shift

    if weight >= bridge_weight + tw
      bridge_weight += tw
      bridge << tw
    else
      # 다리위에 있는 트럭 무게와 새로 올릴 트럭 무게합이 중량 제한보다 작아질 때까지 다리 위 트럭을 뺀다.
      while weight < bridge_weight + tw
        bridge_weight -= bridge.shift
        bridge << 0
        answer += 1
      end
      bridge_weight += tw
      bridge << tw
    end
  end

  # 다리 위 트럭을 모두 뺀다
  while bridge_weight > 0
    answer += 1
    bridge_weight -= bridge.shift
  end
  return answer
end

p solution 2, 10, [7, 4, 5, 6]
p solution 100, 100, [10]
p solution 100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
